x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java

```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+        MultiSearchResponse.Item[] responses = p.getResponses();
+        int nextResponse = 0;
+        for (Sequence sequence : batchToCheck) {
+            boolean leading = true;
+            boolean discarded = false;
+            Timestamp lastLeading = null;
+            Timestamp firstTrailing = null;
+            for (int i = 0; i < criteria.size(); i++) {
+                SequenceCriterion criterion = criteria.get(i);
+                if (criterion.missing()) {
+                    SearchResponse response = responses[nextResponse++].getResponse();
+                    if (discarded) {
+                        continue; // consume all the responses for this sequence even if it's already discarded
+                    }
+                    SearchHit[] hits = response.getHits().getHits();
+                    if (leading) {
+                        if (hits.length == 0) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        lastLeading = lastLeading == null || lastLeading.delta(hitTimestamp) < 0 ? hitTimestamp : lastLeading;
+                    } else if (trailing(i)) {
+                        if (hits.length == 0) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        firstTrailing = firstTrailing == null || firstTrailing.delta(hitTimestamp) > 0 ? hitTimestamp : firstTrailing;
+                    } else {
+                        if (hits.length > 0) {
+                            discarded = true;
+                        }
+                    }
+                } else {
+                    leading = false;
+                }
+            }
+            if (discarded == false) {
+                int lastStage = criteria.size() - 1;
+                if ((firstTrailing == null && lastLeading == null)
+                    || (lastLeading == null && matcher.isMissingEvent(0))
+                    || (firstTrailing == null && matcher.isMissingEvent(lastStage))
+                    || (matcher.isMissingEvent(0) && matcher.isMissingEvent(lastStage) && biggerThanMaxSpan(lastLeading, firstTrailing))
+                    || (matcher.isMissingEvent(0)
+                        && matcher.isMissingEvent(lastStage) == false
+                        && biggerThanMaxSpan(lastLeading, sequence.ordinal().timestamp()))
+                    || (matcher.isMissingEvent(0) == false
+                        && matcher.isMissingEvent(lastStage)
+                        && biggerThanMaxSpan(sequence.startOrdinal().timestamp(), firstTrailing))
+
+                ) {
+                    matcher.addToCompleted(sequence);
+                }
+            }
+        }
+        checkMissingEvents(next, listener);
+    }
+
+    private boolean biggerThanMaxSpan(Timestamp from, Timestamp to) {
```

Comment:
`SequenceMatcher.exceedsMasSpan` is doing almost the same thing. My suggestion is to externalize this behavior and make it generic so that both places can use it. Also, in SequenceMatcher the `delta` is used which seems the better choice.